File: packages/nerva-js/tests/context.test.ts
import { describe, it, expect } from "vitest";
import {
  ExecContext,
  createPermissions,
  TokenUsage,
  InMemoryStreamSink,
  SCOPE_VALUES,
} from "../src/context.js";
import type { Scope } from "../src/context.js";

// ---------------------------------------------------------------------------
// ExecContext.create()
// ---------------------------------------------------------------------------

describe("ExecContext.create", () => {
  it("produces a context with non-empty requestId and traceId", () => {
    const ctx = ExecContext.create();
    expect(ctx.requestId).toBeTruthy();
    expect(ctx.traceId).toBeTruthy();
    expect(ctx.requestId).not.toBe(ctx.traceId);
  });

  it("generates hex-only IDs (no dashes from UUID)", () => {
    const ctx = ExecContext.create();
    expect(ctx.requestId).toMatch(/^[0-9a-f]+$/);
    expect(ctx.traceId).toMatch(/^[0-9a-f]+$/);
  });

  it("generates unique IDs on each call", () => {
    const a = ExecContext.create();
    const b = ExecContext.create();
    expect(a.requestId).not.toBe(b.requestId);
    expect(a.traceId).not.toBe(b.traceId);
  });

  it("defaults userId and sessionId to null", () => {
    const ctx = ExecContext.create();
    expect(ctx.userId).toBeNull();
    expect(ctx.sessionId).toBeNull();
  });

  it("defaults memoryScope to 'session'", () => {
    const ctx = ExecContext.create();
    expect(ctx.memoryScope).toBe("session");
  });

  it("respects custom options", () => {
    const perms = createPermissions({ roles: new Set(["admin"]) });
    const ctx = ExecContext.create({
      userId: "u-1",
      sessionId: "s-1",
      permissions: perms,
      memoryScope: "user",
      timeoutSeconds: 10,
    });
    expect(ctx.userId).toBe("u-1");
    expect(ctx.sessionId).toBe("s-1");
    expect(ctx.permissions.hasRole("admin")).toBe(true);
    expect(ctx.memoryScope).toBe("user");
    expect(ctx.timeoutAt).not.toBeNull();
  });

  it("starts with empty spans, events, and zero token usage", () => {
    const ctx = ExecContext.create();
    expect(ctx.spans).toHaveLength(0);
    expect(ctx.events).toHaveLength(0);
    expect(ctx.tokenUsage.totalTokens).toBe(0);
  });

  it("sets stream to null when not provided", () => {
    const ctx = ExecContext.create();
    expect(ctx.stream).toBeNull();
  });

  it("accepts a StreamSink via options", () => {
    const sink = new InMemoryStreamSink();
    const ctx = ExecContext.create({ stream: sink });
    expect(ctx.stream).toBe(sink);
  });

  it("sets timeoutAt to null when no timeout specified", () => {
    const ctx = ExecContext.create();
    expect(ctx.timeoutAt).toBeNull();
  });

  it("handles empty string userId", () => {
    const ctx = ExecContext.create({ userId: "" });
    expect(ctx.userId).toBe("");
  });

  it("handles very long userId without error", () => {
    const longId = "x".repeat(10_000);
    const ctx = ExecContext.create({ userId: longId });
    expect(ctx.userId).toBe(longId);
  });
});

// ---------------------------------------------------------------------------
// ExecContext.child()
// ---------------------------------------------------------------------------

describe("ExecContext.child", () => {
  it("inherits traceId from parent", () => {
    const parent = ExecContext.create({ userId: "u-1", sessionId: "s-1" });
    const child = parent.child("sub-handler");
    expect(child.traceId).toBe(parent.traceId);
  });

  it("gets a fresh requestId different from parent", () => {
    const parent = ExecContext.create();
    const child = parent.child("sub");
    expect(child.requestId).not.toBe(parent.requestId);
  });

  it("inherits permissions from parent", () => {
    const perms = createPermissions({ roles: new Set(["admin"]) });
    const parent = ExecContext.create({ permissions: perms });
    const child = parent.child("sub");
    expect(child.permissions.hasRole("admin")).toBe(true);
  });

  it("inherits userId and sessionId", () => {
    const parent = ExecContext.create({ userId: "u-1", sessionId: "s-1" });
    const child = parent.child("sub");
    expect(child.userId).toBe("u-1");
    expect(child.sessionId).toBe("s-1");
  });

  it("inherits memoryScope", () => {
    const parent = ExecContext.create({ memoryScope: "global" });
    const child = parent.child("sub");
    expect(child.memoryScope).toBe("global");
  });

  it("shares cancellation controller with parent", () => {
    const parent = ExecContext.create();
    const child = parent.child("sub");
    parent.cancel();
    expect(child.isCancelled()).toBe(true);
  });

  it("starts with a root span named after handlerName", () => {
    const parent = ExecContext.create();
    const child = parent.child("my-handler");
    expect(child.spans).toHaveLength(1);
    expect(child.spans[0]!.name).toBe("my-handler");
    expect(child.spans[0]!.parentId).toBe(parent.requestId);
  });

  it("starts with empty events and zero token usage", () => {
    const parent = ExecContext.create();
    const child = parent.child("sub");
    expect(child.events).toHaveLength(0);
    expect(child.tokenUsage.totalTokens).toBe(0);
  });

  it("inherits metadata as a shallow copy", () => {
    const parent = ExecContext.create();
    parent.metadata["key"] = "val";
    const child = parent.child("sub");
    expect(child.metadata["key"]).toBe("val");
    // mutation on child does not affect parent
    child.metadata["key"] = "changed";
    expect(parent.metadata["key"]).toBe("val");
  });
});

// ---------------------------------------------------------------------------
// Permissions
// ---------------------------------------------------------------------------

describe("createPermissions", () => {
  it("defaults to unrestricted (null allowlists)", () => {
    const p = createPermissions();
    expect(p.canUseTool("anything")).toBe(true);
    expect(p.canUseAgent("anything")).toBe(true);
  });

  it("canUseTool returns false for non-allowed tools", () => {
    const p = createPermissions({ allowedTools: new Set(["search"]) });
    expect(p.canUseTool("search")).toBe(true);
    expect(p.canUseTool("delete")).toBe(false);
  });

  it("canUseAgent returns false for non-allowed agents", () => {
    const p = createPermissions({ allowedAgents: new Set(["helper"]) });
    expect(p.canUseAgent("helper")).toBe(true);
    expect(p.canUseAgent("destroyer")).toBe(false);
  });

  it("hasRole checks role membership", () => {
    const p = createPermissions({ roles: new Set(["admin", "user"]) });
    expect(p.hasRole("admin")).toBe(true);
    expect(p.hasRole("superadmin")).toBe(false);
  });

  it("empty allowedTools set blocks all tools", () => {
    const p = createPermissions({ allowedTools: new Set<string>() });
    expect(p.canUseTool("anything")).toBe(false);
  });

  it("empty allowedAgents set blocks all agents", () => {
    const p = createPermissions({ allowedAgents: new Set<string>() });
    expect(p.canUseAgent("anything")).toBe(false);
  });

  it("empty roles set means hasRole always false", () => {
    const p = createPermissions();
    expect(p.hasRole("admin")).toBe(false);
  });

  it("is frozen (immutable)", () => {
    const p = createPermissions();
    expect(Object.isFrozen(p)).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// TokenUsage
// ---------------------------------------------------------------------------

describe("TokenUsage", () => {
  it("defaults all fields to 0", () => {
    const u = new TokenUsage();
    expect(u.promptTokens).toBe(0);
    expect(u.completionTokens).toBe(0);
    expect(u.totalTokens).toBe(0);
    expect(u.costUsd).toBe(0);
  });

  it("add() returns a new instance with summed fields", () => {
    const a = new TokenUsage(10, 20, 30, 0.5);
    const b = new TokenUsage(5, 10, 15, 0.25);
    const result = a.add(b);
    expect(result.promptTokens).toBe(15);
    expect(result.completionTokens).toBe(30);
    expect(result.totalTokens).toBe(45);
    expect(result.costUsd).toBeCloseTo(0.75);
  });

  it("add() does not mutate the original", () => {
    const a = new TokenUsage(10, 20, 30, 0.5);
    const b = new TokenUsage(5, 10, 15, 0.25);
    a.add(b);
    expect(a.promptTokens).toBe(10);
    expect(a.completionTokens).toBe(20);
  });

  it("add() with two zero-value instances returns zero", () => {
    const a = new TokenUsage();
    const b = new TokenUsage();
    const result = a.add(b);
    expect(result.totalTokens).toBe(0);
    expect(result.costUsd).toBe(0);
  });
});

// ---------------------------------------------------------------------------
// InMemoryStreamSink
// ---------------------------------------------------------------------------

describe("InMemoryStreamSink", () => {
  it("collects pushed chunks in order", async () => {
    const sink = new InMemoryStreamSink();
    await sink.push("hello ");
    await sink.push("world");
    expect(sink.chunks).toEqual(["hello ", "world"]);
  });

  it("starts not closed", () => {
    const sink = new InMemoryStreamSink();
    expect(sink.closed).toBe(false);
  });

  it("sets closed to true after close()", async () => {
    const sink = new InMemoryStreamSink();
    await sink.close();
    expect(sink.closed).toBe(true);
  });

  it("throws on push after close", async () => {
    const sink = new InMemoryStreamSink();
    await sink.close();
    await expect(sink.push("x")).rejects.toThrow("Cannot push to a closed StreamSink");
  });

  it("throws on double close", async () => {
    const sink = new InMemoryStreamSink();
    await sink.close();
    await expect(sink.close()).rejects.toThrow("StreamSink is already closed");
  });

  it("handles empty string push", async () => {
    const sink = new InMemoryStreamSink();
    await sink.push("");
    expect(sink.chunks).toEqual([""]);
  });

  it("handles unicode content", async () => {
    const sink = new InMemoryStreamSink();
    await sink.push("\u{1F600} hello \u4E16\u754C");
    expect(sink.chunks[0]).toContain("\u{1F600}");
  });
});

// ---------------------------------------------------------------------------
// isTimedOut / isCancelled
// ---------------------------------------------------------------------------

describe("ExecContext timeout and cancellation", () => {
  it("isTimedOut returns false when no timeout set", () => {
    const ctx = ExecContext.create();
    expect(ctx.isTimedOut()).toBe(false);
  });

  it("isTimedOut returns false when timeout is far in the future", () => {
    const ctx = ExecContext.create({ timeoutSeconds: 3600 });
    expect(ctx.isTimedOut()).toBe(false);
  });

  it("isTimedOut returns true when timeout is in the past", () => {
    const ctx = ExecContext.create({ timeoutSeconds: -1 });
    expect(ctx.isTimedOut()).toBe(true);
  });

  it("isCancelled returns false initially", () => {
    const ctx = ExecContext.create();
    expect(ctx.isCancelled()).toBe(false);
  });

  it("isCancelled returns true after cancel()", () => {
    const ctx = ExecContext.create();
    ctx.cancel();
    expect(ctx.isCancelled()).toBe(true);
  });

  it("cancelSignal is an AbortSignal", () => {
    const ctx = ExecContext.create();
    expect(ctx.cancelSignal).toBeInstanceOf(AbortSignal);
  });

  it("elapsedSeconds returns a non-negative number", () => {
    const ctx = ExecContext.create();
    expect(ctx.elapsedSeconds()).toBeGreaterThanOrEqual(0);
  });
});

// ---------------------------------------------------------------------------
// Span and Event helpers
// ---------------------------------------------------------------------------

describe("ExecContext span and event helpers", () => {
  it("addSpan appends a span with the given name", () => {
    const ctx = ExecContext.create();
    const span = ctx.addSpan("llm.call");
    expect(span.name).toBe("llm.call");
    expect(span.endedAt).toBeNull();
    expect(ctx.spans).toHaveLength(1);
  });

  it("addEvent appends an event with the given name", () => {
    const ctx = ExecContext.create();
    const event = ctx.addEvent("policy.denied", { target: "deploy" });
    expect(event.name).toBe("policy.denied");
    expect(event.attributes["target"]).toBe("deploy");
    expect(ctx.events).toHaveLength(1);
  });

  it("recordTokens accumulates usage immutably", () => {
    const ctx = ExecContext.create();
    const usage = new TokenUsage(100, 50, 150, 0.01);
    ctx.recordTokens(usage);
    expect(ctx.tokenUsage.totalTokens).toBe(150);
    ctx.recordTokens(usage);
    expect(ctx.tokenUsage.totalTokens).toBe(300);
  });
});

// ---------------------------------------------------------------------------
// Scope values
// ---------------------------------------------------------------------------

describe("SCOPE_VALUES", () => {
  it("contains all four valid scopes", () => {
    expect(SCOPE_VALUES).toContain("user");
    expect(SCOPE_VALUES).toContain("session");
    expect(SCOPE_VALUES).toContain("agent");
    expect(SCOPE_VALUES).toContain("global");
    expect(SCOPE_VALUES).toHaveLength(4);
  });
});

File: packages/nerva-go/context/context_test.go
package context

import (
	"sync"
	"testing"
	"time"
)

// --- NewContext ---

func TestNewContextDefaultValues(t *testing.T) {
	ctx := NewContext()

	if ctx.RequestID == "" {
		t.Fatal("expected non-empty RequestID")
	}
	if ctx.TraceID == "" {
		t.Fatal("expected non-empty TraceID")
	}
	if ctx.Depth != 0 {
		t.Fatalf("expected Depth 0, got %d", ctx.Depth)
	}
	if ctx.MemoryScope != DefaultMemoryScope {
		t.Fatalf("expected MemoryScope %q, got %q", DefaultMemoryScope, ctx.MemoryScope)
	}
	if ctx.TimeoutAt != nil {
		t.Fatal("expected nil TimeoutAt without timeout option")
	}
	if ctx.TokenUsage == nil {
		t.Fatal("expected non-nil TokenUsage")
	}
	if ctx.Metadata == nil {
		t.Fatal("expected non-nil Metadata map")
	}
	if len(ctx.Spans()) != 0 {
		t.Fatalf("expected 0 spans, got %d", len(ctx.Spans()))
	}
	if len(ctx.Events()) != 0 {
		t.Fatalf("expected 0 events, got %d", len(ctx.Events()))
	}
}

func TestNewContextWithOptions(t *testing.T) {
	perms := Permissions{
		Roles: map[string]bool{"admin": true},
	}
	ctx := NewContext(
		WithUserID("u1"),
		WithSessionID("s1"),
		WithPermissions(perms),
		WithMemoryScope(ScopeUser),
		WithTimeout(5.0),
	)

	if ctx.UserID != "u1" {
		t.Fatalf("expected UserID u1, got %q", ctx.UserID)
	}
	if ctx.SessionID != "s1" {
		t.Fatalf("expected SessionID s1, got %q", ctx.SessionID)
	}
	if !ctx.Permissions.HasRole("admin") {
		t.Fatal("expected admin role")
	}
	if ctx.MemoryScope != ScopeUser {
		t.Fatalf("expected ScopeUser, got %q", ctx.MemoryScope)
	}
	if ctx.TimeoutAt == nil {
		t.Fatal("expected non-nil TimeoutAt with timeout option")
	}
}

func TestNewContextUniqueIDs(t *testing.T) {
	a := NewContext()
	b := NewContext()
	if a.RequestID == b.RequestID {
		t.Fatal("expected unique RequestIDs")
	}
	if a.TraceID == b.TraceID {
		t.Fatal("expected unique TraceIDs")
	}
}

// --- Child ---

func TestChildInheritsParentFields(t *testing.T) {
	parent := NewContext(
		WithUserID("u1"),
		WithSessionID("s1"),
		WithMemoryScope(ScopeAgent),
		WithTimeout(10.0),
	)
	parent.Metadata["key"] = "val"

	child := parent.Child("test-handler")

	if child.TraceID != parent.TraceID {
		t.Fatal("child should inherit TraceID")
	}
	if child.UserID != parent.UserID {
		t.Fatal("child should inherit UserID")
	}
	if child.SessionID != parent.SessionID {
		t.Fatal("child should inherit SessionID")
	}
	if child.MemoryScope != parent.MemoryScope {
		t.Fatal("child should inherit MemoryScope")
	}
	if child.TimeoutAt == nil || parent.TimeoutAt == nil {
		t.Fatal("both should have TimeoutAt")
	}
	if *child.TimeoutAt != *parent.TimeoutAt {
		t.Fatal("child should inherit TimeoutAt")
	}
	if child.Depth != parent.Depth+1 {
		t.Fatalf("expected child depth %d, got %d", parent.Depth+1, child.Depth)
	}
	if child.RequestID == parent.RequestID {
		t.Fatal("child should have a fresh RequestID")
	}
	if child.Metadata["key"] != "val" {
		t.Fatal("child should copy metadata")
	}
}

func TestChildMetadataIsolation(t *testing.T) {
	parent := NewContext()
	parent.Metadata["shared"] = "yes"

	child := parent.Child("h")
	child.Metadata["child_only"] = "true"

	if _, ok := parent.Metadata["child_only"]; ok {
		t.Fatal("child metadata mutation must not affect parent")
	}
}

func TestChildHasRootSpan(t *testing.T) {
	parent := NewContext()
	child := parent.Child("my-handler")

	spans := child.Spans()
	if len(spans) != 1 {
		t.Fatalf("expected 1 root span, got %d", len(spans))
	}
	if spans[0].Name != "my-handler" {
		t.Fatalf("expected span name 'my-handler', got %q", spans[0].Name)
	}
}

func TestChildTokenUsageIndependent(t *testing.T) {
	parent := NewContext()
	child := parent.Child("h")

	child.TokenUsage.Accumulate(&TokenUsage{PromptTokens: 10, TotalTokens: 10})

	if parent.TokenUsage.TotalTokens != 0 {
		t.Fatal("child token usage should not affect parent")
	}
}

// --- Permissions ---

func TestCanUseToolUnrestricted(t *testing.T) {
	p := Permissions{AllowedTools: nil}
	if !p.CanUseTool("anything") {
		t.Fatal("nil AllowedTools should allow all")
	}
}

func TestCanUseToolRestricted(t *testing.T) {
	allowed := map[string]bool{"read": true}
	p := Permissions{AllowedTools: &allowed}

	if !p.CanUseTool("read") {
		t.Fatal("expected read to be allowed")
	}
	if p.CanUseTool("write") {
		t.Fatal("expected write to be denied")
	}
}

func TestCanUseToolEmptySet(t *testing.T) {
	empty := map[string]bool{}
	p := Permissions{AllowedTools: &empty}
	if p.CanUseTool("anything") {
		t.Fatal("empty allowed set should deny all")
	}
}

func TestCanUseAgentUnrestricted(t *testing.T) {
	p := Permissions{AllowedAgents: nil}
	if !p.CanUseAgent("any") {
		t.Fatal("nil AllowedAgents should allow all")
	}
}

func TestCanUseAgentRestricted(t *testing.T) {
	allowed := map[string]bool{"agent-a": true}
	p := Permissions{AllowedAgents: &allowed}

	if !p.CanUseAgent("agent-a") {
		t.Fatal("expected agent-a to be allowed")
	}
	if p.CanUseAgent("agent-b") {
		t.Fatal("expected agent-b to be denied")
	}
}

func TestHasRoleNilRoles(t *testing.T) {
	p := Permissions{}
	if p.HasRole("admin") {
		t.Fatal("nil Roles map should return false")
	}
}

func TestHasRole(t *testing.T) {
	p := Permissions{Roles: map[string]bool{"admin": true}}
	if !p.HasRole("admin") {
		t.Fatal("expected true for admin")
	}
	if p.HasRole("user") {
		t.Fatal("expected false for user")
	}
}

// --- TokenUsage ---

func TestTokenUsageAdd(t *testing.T) {
	a := &TokenUsage{PromptTokens: 10, CompletionTokens: 5, TotalTokens: 15, CostUSD: 0.01}
	b := &TokenUsage{PromptTokens: 20, CompletionTokens: 10, TotalTokens: 30, CostUSD: 0.02}

	result := a.Add(b)

	if result.PromptTokens != 30 {
		t.Fatalf("expected 30 prompt tokens, got %d", result.PromptTokens)
	}
	if result.TotalTokens != 45 {
		t.Fatalf("expected 45 total tokens, got %d", result.TotalTokens)
	}
	if result.CostUSD != 0.03 {
		t.Fatalf("expected 0.03 cost, got %f", result.CostUSD)
	}
}

func TestTokenUsageAccumulate(t *testing.T) {
	a := &TokenUsage{PromptTokens: 10, TotalTokens: 10}
	b := &TokenUsage{PromptTokens: 5, TotalTokens: 5}

	a.Accumulate(b)

	if a.PromptTokens != 15 {
		t.Fatalf("expected 15, got %d", a.PromptTokens)
	}
	if a.TotalTokens != 15 {
		t.Fatalf("expected 15, got %d", a.TotalTokens)
	}
}

func TestTokenUsageConcurrentAccumulate(t *testing.T) {
	target := &TokenUsage{}
	var wg sync.WaitGroup
	iterations := 100

	for i := 0; i < iterations; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			target.Accumulate(&TokenUsage{TotalTokens: 1})
		}()
	}
	wg.Wait()

	if target.TotalTokens != iterations {
		t.Fatalf("expected %d, got %d (race condition?)", iterations, target.TotalTokens)
	}
}

// --- Timeout / Cancel ---

func TestIsTimedOutNoTimeout(t *testing.T) {
	ctx := NewContext()
	if ctx.IsTimedOut() {
		t.Fatal("should not be timed out without timeout")
	}
}

func TestIsTimedOutExpired(t *testing.T) {
	ctx := NewContext(WithTimeout(0.001)) // 1ms
	time.Sleep(5 * time.Millisecond)
	if !ctx.IsTimedOut() {
		t.Fatal("should be timed out")
	}
}

func TestCancelAndIsCancelled(t *testing.T) {
	ctx := NewContext()
	if ctx.IsCancelled() {
		t.Fatal("should not be cancelled yet")
	}
	ctx.Cancel()
	if !ctx.IsCancelled() {
		t.Fatal("should be cancelled after Cancel()")
	}
}

func TestCancelPropagatesToChild(t *testing.T) {
	parent := NewContext()
	child := parent.Child("h")

	parent.Cancel()

	if !child.IsCancelled() {
		t.Fatal("child should be cancelled when parent is cancelled")
	}
}

// --- Spans / Events ---

func TestAddSpan(t *testing.T) {
	ctx := NewContext()
	span := ctx.AddSpan("my-span")

	if span.Name != "my-span" {
		t.Fatalf("expected name 'my-span', got %q", span.Name)
	}
	if span.SpanID == "" {
		t.Fatal("expected non-empty SpanID")
	}
	if span.ParentID != ctx.RequestID {
		t.Fatal("expected ParentID == RequestID")
	}

	spans := ctx.Spans()
	if len(spans) != 1 {
		t.Fatalf("expected 1 span, got %d", len(spans))
	}
}

func TestAddEvent(t *testing.T) {
	ctx := NewContext()
	event := ctx.AddEvent("test-event", map[string]string{"k": "v"})

	if event.Name != "test-event" {
		t.Fatalf("expected name 'test-event', got %q", event.Name)
	}
	if event.Attributes["k"] != "v" {
		t.Fatal("expected attribute k=v")
	}

	events := ctx.Events()
	if len(events) != 1 {
		t.Fatalf("expected 1 event, got %d", len(events))
	}
}

func TestAddEventNilAttributes(t *testing.T) {
	ctx := NewContext()
	event := ctx.AddEvent("e", nil)

	if event.Attributes == nil {
		t.Fatal("nil attributes should be replaced with empty map")
	}
}

func TestSpansReturnsCopy(t *testing.T) {
	ctx := NewContext()
	ctx.AddSpan("s1")

	spans := ctx.Spans()
	spans = append(spans, Span{Name: "injected"})

	if len(ctx.Spans()) != 1 {
		t.Fatal("Spans() must return a copy, not a reference")
	}
}

func TestEventsReturnsCopy(t *testing.T) {
	ctx := NewContext()
	ctx.AddEvent("e1", nil)

	events := ctx.Events()
	events = append(events, Event{Name: "injected"})

	if len(ctx.Events()) != 1 {
		t.Fatal("Events() must return a copy, not a reference")
	}
}

func TestConcurrentSpanAndEventAdds(t *testing.T) {
	ctx := NewContext()
	var wg sync.WaitGroup
	count := 50

	for i := 0; i < count; i++ {
		wg.Add(2)
		go func() {
			defer wg.Done()
			ctx.AddSpan("s")
		}()
		go func() {
			defer wg.Done()
			ctx.AddEvent("e", nil)
		}()
	}
	wg.Wait()

	if len(ctx.Spans()) != count {
		t.Fatalf("expected %d spans, got %d", count, len(ctx.Spans()))
	}
	if len(ctx.Events()) != count {
		t.Fatalf("expected %d events, got %d", count, len(ctx.Events()))
	}
}

// --- RecordTokens ---

func TestRecordTokens(t *testing.T) {
	ctx := NewContext()
	ctx.RecordTokens(&TokenUsage{PromptTokens: 10, TotalTokens: 10})
	ctx.RecordTokens(&TokenUsage{PromptTokens: 5, TotalTokens: 5})

	if ctx.TokenUsage.TotalTokens != 15 {
		t.Fatalf("expected 15, got %d", ctx.TokenUsage.TotalTokens)
	}
}

// --- ElapsedSeconds ---

func TestElapsedSecondsPositive(t *testing.T) {
	ctx := NewContext()
	time.Sleep(2 * time.Millisecond)
	elapsed := ctx.ElapsedSeconds()
	if elapsed <= 0 {
		t.Fatalf("expected positive elapsed, got %f", elapsed)
	}
}

// --- Context() ---

func TestContextReturnsStdContext(t *testing.T) {
	ctx := NewContext()
	stdCtx := ctx.Context()
	if stdCtx == nil {
		t.Fatal("expected non-nil context.Context")
	}
}

// --- Edge cases ---

func TestNewContextEmptyUserID(t *testing.T) {
	ctx := NewContext(WithUserID(""))
	if ctx.UserID != "" {
		t.Fatal("expected empty UserID")
	}
}

func TestNewContextZeroTimeout(t *testing.T) {
	ctx := NewContext(WithTimeout(0))
	if ctx.TimeoutAt != nil {
		t.Fatal("zero timeout should not set TimeoutAt")
	}
}

func TestNewContextNegativeTimeout(t *testing.T) {
	ctx := NewContext(WithTimeout(-1))
	if ctx.TimeoutAt != nil {
		t.Fatal("negative timeout should not set TimeoutAt")
	}
}
